### 通函数

numpy提供熟悉的数学函数，例如sin,cos和exp。在numpy中，这些被称为“通函数”（ufunc）。在numpy中，这些函数在数组上按元素进行运算，产生一个数组作为输出。

In [1]:
import numpy as np

In [2]:
x = np.arange(3)

In [3]:
x

array([0, 1, 2])

In [4]:
np.exp(x) #以自然常数e为底的指数

array([1.        , 2.71828183, 7.3890561 ])

In [5]:
np.sqrt(x) #开平方

array([0.        , 1.        , 1.41421356])

In [6]:
C = np.array([2,-1,4])

In [7]:
np.add(x,C)

array([2, 0, 6])

#### np.all()

In [8]:
np.all([[True,False],[True,True]])

False

In [9]:
np.all([[True,False],[True,True]],axis=0)

array([ True, False])

In [10]:
np.all([-1,4,5])

True

In [11]:
np.all([1.0,np.nan])

True

In [12]:
np.all([0,1,3]) #元素值为0才对应false

False

In [13]:
a = np.array([1,2,3])
b = np.array([1,2,4])

In [14]:
a==b

array([ True,  True, False])

In [15]:
#为了判断两个数组是否完整一致，可以这样使用
(a==b).all()

False

In [16]:
#或者这样
np.all((a==b))

False

#### np.any()

和all正好相反

In [17]:
np.any([[True,False],[False,False]]) #有一个元素对应True，则返回的结果就是True

True

In [18]:
np.any([[True,False],[False,False]],axis=0)

array([ True, False])

In [19]:
(a==b).any()

True

#### np.apply_along_axis(func1d, axis, arr, *args, **kwargs)

将一个函数应用于一个一维的数组上（由轴决定）

In [20]:
def my_func(a):
    """计算头尾两个元素之和的平均值"""
    return (a[0]+a[-1])*0.5

In [21]:
b = np.array([[1,2,3],[4,5,6],[7,8,9]])

In [22]:
np.apply_along_axis(my_func,0,b)

array([4., 5., 6.])

In [23]:
np.apply_along_axis(my_func,1,b)

array([2., 5., 8.])

还可以使用一些默认的函数，比如sorted进行排序

In [24]:
b = np.array([[8,1,7], [4,3,9], [5,2,6]])

In [25]:
np.apply_along_axis(sorted,1,b)

array([[1, 7, 8],
       [3, 4, 9],
       [2, 5, 6]])

#### np.argmax(a, axis=None, out=None)

返回在一个轴上最大值的索引

In [26]:
a = np.arange(6).reshape(2,3) + 10

In [27]:
a

array([[10, 11, 12],
       [13, 14, 15]])

In [28]:
np.argmax(a)

5

In [29]:
np.argmax(a,axis=0)

array([1, 1, 1], dtype=int64)

In [30]:
np.argmax(a,axis=1)

array([2, 2], dtype=int64)

In [31]:
b = np.arange(6)

In [32]:
b[1] = 5

In [33]:
b

array([0, 5, 2, 3, 4, 5])

In [34]:
np.argmax(b) #返回第一个最大值的索引

1

#### np.argmin(a, axis=None, out=None)

和argmax相反

#### np.argsort(a, axis=-1, kind=None, order=None)

返回指定轴上一维数组排序后索引组成的数组

In [35]:
x = np.array([3,1,2])

In [36]:
np.argsort(x)

array([1, 2, 0], dtype=int64)

In [37]:
x = np.array([[0,5,8],[4,3,-1],[9,10,20]])

In [38]:
ind = np.argsort(x,axis=0)

In [39]:
ind

array([[0, 1, 1],
       [1, 0, 0],
       [2, 2, 2]], dtype=int64)

In [40]:
np.sort(x,axis=0)

array([[ 0,  3, -1],
       [ 4,  5,  8],
       [ 9, 10, 20]])

In [41]:
np.take_along_axis(x,ind,axis=0)

array([[ 0,  3, -1],
       [ 4,  5,  8],
       [ 9, 10, 20]])

#### np.average(a, axis=None, weights=None, returned=False)

计算指定轴上的加权平均值

In [42]:
a = np.arange(15).reshape(3,5)

In [43]:
a

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

In [44]:
np.average(a,axis=0,weights=(10,5,1))

array([2.1875, 3.1875, 4.1875, 5.1875, 6.1875])

[ 2.1875,  3.1875,  4.1875,  5.1875,  6.1875]  =  ( 10*[ 0, 1, 2, 3, 4] + 5*[ 5, 6, 7, 8, 9] + 1*[10, 11, 12, 13, 14]) / (10+5+1)

#### np.around(),np.floor(),np.ceil()

#### np.around()

np.around 返回四舍五入后的值，可指定精度

around(a, decimals=0, out=None),
decimals:要舍入的小数位数，默认值为0，如果为负，整数将四舍五入到小数点左侧的位置

In [46]:
n = np.array([-0.746, 4.6, 9.4, 7.447, 10.455, 11.555])

In [47]:
around1 = np.around(n)

In [48]:
around1

array([-1.,  5.,  9.,  7., 10., 12.])

In [49]:
around2 = np.around(n,decimals=1)

In [50]:
around2

array([-0.7,  4.6,  9.4,  7.4, 10.5, 11.6])

In [51]:
around3 = np.around(n,decimals=-1)

In [52]:
around3

array([-0.,  0., 10., 10., 10., 10.])

#### np.floor()

返回不大于输入参数的最大整数。即对于输入值x，将返回最大的整数i,使得i<=x。注意，在python中，向下取整总是从0舍入

In [55]:
n = np.array([-1.7, -2.5, -0.2, 0.6, 1.2, 2.7, 11])

In [56]:
np.floor(n)

array([-2., -3., -1.,  0.,  1.,  2., 11.])

#### np.ceil()

返回输入值的上限，即对于输入x,返回最小的整数i,使得i>=x

In [57]:
n = np.array([-1.7, -2.5, -0.2, 0.6, 1.2, 2.7, 11])

In [58]:
np.ceil(n)

array([-1., -2., -0.,  1.,  2.,  3., 11.])

#### np.where(condition[, x, y])

根据condition从x和y中选择元素，当为True时，选x,否则选y

In [59]:
n = np.random.random([2,3])

In [60]:
n

array([[0.80586919, 0.48082985, 0.84183877],
       [0.84655569, 0.73433131, 0.49120108]])

In [61]:
np.where(n>0.5,n,0)

array([[0.80586919, 0.        , 0.84183877],
       [0.84655569, 0.73433131, 0.        ]])